File: Solutions/cpp/q1.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> findMinCostServer(int &n1, int &n2, vector<int> &serverLoad, vector<int> &serverCapacity) {
    // Never change the input array in an interview. Bad practice !!. Hence creating a copy :)
    vector<int> copy_of_capacity = serverCapacity;
    vector<int> copy_of_load = serverLoad;

    sort(copy_of_capacity.begin(), copy_of_capacity.end(), greater<int> ());

    sort(copy_of_load.begin(), copy_of_load.end());

    map<int, vector<int>> hashMap;
    map<int, vector<int>> mapIdx;
    for(int i = 0; i < n1; i++) {
        hashMap[copy_of_load[i]].push_back(copy_of_capacity[i]);
        mapIdx[serverLoad[i]].push_back(i);
    }

    vector<int> ans(n2);

    for(auto it : hashMap) {
        vector<int> elements = it.second;
        // Since problem statement has mentioned to return lexicographically smallest.
        sort(elements.begin(), elements.end());

        vector<int> idx = mapIdx[it.first];
        for(int k = 0; k < idx.size(); k++) {
            ans[idx[k]] = elements[k];
        }
    }

    return ans;

}

int main() {
    int n1, n2;
    cin >> n1 >> n2;

    vector<int> serverLoad;
    for(int i = 0; i < n1; i++) {
        int d; cin >> d;
        serverLoad.push_back(d);
    }    

    vector<int> serverCapacity;
    for(int i = 0; i < n2; i++) {
        int d; cin >> d;
        serverCapacity.push_back(d);
    }  

    vector<int> sequence = findMinCostServer(n1, n2, serverLoad, serverCapacity);
    for(int i = 0; i < n2; i++) {
        cout << sequence[i] << " ";
    }

    cout << "\n";

    return 0;
}
 

File: Solutions/java/q1.java
import java.util.*;

public class Main {

    public static List<Integer> findMinCostServer(int n1, int n2, List<Integer> serverLoad, List<Integer> serverCapacity) {
        // Never change the input array in an interview. Bad practice !!. Hence creating a copy :)
        List<Integer> copyOfCapacity = new ArrayList<>(serverCapacity);
        List<Integer> copyOfLoad = new ArrayList<>(serverLoad);

        Collections.sort(copyOfCapacity, Collections.reverseOrder());
        Collections.sort(copyOfLoad);

        Map<Integer, List<Integer>> hashMap = new HashMap<>();
        Map<Integer, List<Integer>> mapIdx = new HashMap<>();
        for (int i = 0; i < n1; i++) {
            hashMap.putIfAbsent(copyOfLoad.get(i), new ArrayList<>());
            hashMap.get(copyOfLoad.get(i)).add(copyOfCapacity.get(i));

            mapIdx.putIfAbsent(serverLoad.get(i), new ArrayList<>());
            mapIdx.get(serverLoad.get(i)).add(i);
        }

        List<Integer> ans = new ArrayList<>(Collections.nCopies(n2, 0));

        for (Map.Entry<Integer, List<Integer>> entry : hashMap.entrySet()) {
            List<Integer> elements = entry.getValue();
            // Since problem statement has mentioned to return lexicographically smallest.
            Collections.sort(elements);

            List<Integer> idx = mapIdx.get(entry.getKey());
            for (int k = 0; k < idx.size(); k++) {
                ans.set(idx.get(k), elements.get(k));
            }
        }

        return ans;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n1 = scanner.nextInt();
        int n2 = scanner.nextInt();

        List<Integer> serverLoad = new ArrayList<>();
        for (int i = 0; i < n1; i++) {
            serverLoad.add(scanner.nextInt());
        }

        List<Integer> serverCapacity = new ArrayList<>();
        for (int i = 0; i < n2; i++) {
            serverCapacity.add(scanner.nextInt());
        }

        List<Integer> sequence = findMinCostServer(n1, n2, serverLoad, serverCapacity);
        for (int i = 0; i < n2; i++) {
            System.out.print(sequence.get(i) + " ");
        }

        System.out.println();
    }
}
